sound/soc/sof/intel/hda-dai.c

```diff
@@ -510,6 +510,7 @@ int sdw_hda_dai_hw_params(struct snd_pcm_substream *substream,
 		if (ch_maps->cpu == cpu_dai_id)
 			ch_mask |= ch_maps->ch_mask;
 	}
+	ch_mask = GENMASK(hweight_long(ch_mask) - 1, 0);
```

Comment:
that seems weird, I don't understand this code. It doesn't really align with the comment that talks about "lowest channels",

Reply:
> that seems weird, I don't understand this code. It doesn't really align with the comment that talks about "lowest channels",

What I did is to calculate how many bits are set in ch_mask, and use the lowest N bits in ch_mask. In other words, move those bits to the lowest bits. For example, 0xC -> 0x3. However, I am not sure if 0xA -> 0x3 is correct or not. Or should be 0xA -> 0x5.

Reply:
> Not following @bardliao. If we are using fw-based aggregation and there is a single amp per link, ch_mask should be 0x3 after the OR operation. I am not sure in what cases we would have a mask that's larger than 0x3 ?

Are you suggesting we set ch_mask = 0x3 in link ch maps? I.e. Set ch_maps->ch_mask = 0x3 here https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L628. I thought about it, but it seems can't represent we have 4 separated channels in ASoC level. So, my solution is that we still set 0xc and 0x3 in link ch maps, and convert the ch_map to 0x3 in DAI level, then set the ch_map via hdac_bus_eml_sdw_map_stream_ch().

Reply:
Oh man, this is giving me a migraine.... 
I can't recall what we planned on doing with those ch_maps.

````c
		/*
		 * construct cpu channel mask by combining ch_mask of each
		 * codec which maps to the cpu.
		 * see
		 *	soc.h :: [dai_link->ch_maps Image sample]
		 */
		for_each_rtd_ch_maps(rtd, j, ch_maps)
			if (ch_maps->cpu == i)
				ch_mask |= ch_maps->ch_mask;

	        /* fixup cpu channel number */
		if (ch_mask)
			soc_pcm_codec_params_fixup(&tmp_params, ch_mask);
````

What this does is identify the number of channels handled by the cpu_dai. It doesn't matter if the mask is 0xC or 0x3, in the end the number of channels set for the dai will be 2, and we don't need to fix it up in the cpu_dai hw_params, do we?

What happens if we completely remove all this code and take what soc-pcm.c determined already?

Reply:
> What happens if we completely remove all this code and take what soc-pcm.c determined already?

Yes, it works, and updated the commit.